### PR TITLE
bpo-16845: Validate the category of warnings.simplefilter sooner

### LINKED ISSUE #61049: warnings.simplefilter should validate input
BPO | [16845](https://bugs.python.org/issue16845)
--- | :---
Nosy | @berkerpeksag, @seberg, @vajrasky, @nanjekyejoannah, @Bonifacio2, @DanielNoord
PRs | <li>python/cpython#26696</li><li>python/cpython#31983</li>
Files | <li>[add_assert_in_simplefilter.patch](https://bugs.python.org/file32131/add_assert_in_simplefilter.patch "Uploaded as text/plain at 2013-10-15.08:58:30 by @vajrasky")</li><li>[issue16845.diff](https://bugs.python.org/file35773/issue16845.diff "Uploaded as text/plain at 2014-06-24.19:09:44 by @berkerpeksag")</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2013-01-03.00:42:06.501>
labels = ['easy', 'type-feature', 'library', '3.10']
title = 'warnings.simplefilter should validate input'
updated_at = <Date 2022-03-18.19:52:15.032>
user = 'https://github.com/seberg'
```

bugs.python.org fields:
```python
activity = <Date 2022-03-18.19:52:15.032>
actor = 'danielnoord'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2013-01-03.00:42:06.501>
creator = 'seberg'
dependencies = []
files = ['32131', '35773']
hgrepos = []
issue_num = 16845
keywords = ['patch', 'easy']
message_count = 5.0
messages = ['178867', '199985', '221480', '395747', '415522']
nosy_count = 7.0
nosy_names = ['python-dev', 'berker.peksag', 'seberg', 'vajrasky', 'nanjekyejoannah', 'Bonifacio2', 'danielnoord']
pr_nums = ['26696', '31983']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue16845'
versions = ['Python 3.10']
```

</p></details>

### PR DESCRIPTION
**Pull request by DanielNoord**
This is a submit of the patch provided by @ berkerpeksag in [bpo-16845](https://bugs.python.org/issue16845). I added one additional test case.
_Not sure if I should ping them here._

It supersedes https://github.com/python/cpython/pull/26696 which did not received any interaction after the initial push and has now been marked stale.

https://bugs.python.org/issue16845

Closes #61049

### CODE REVIEW

File: Lib/unittest/case.py [hunk 1]
```diff
@@ -274,7 +274,11 @@ def __enter__(self):
                 v.__warningregistry__ = {}
         self.warnings_manager = warnings.catch_warnings(record=True)
         self.warnings = self.warnings_manager.__enter__()
-        warnings.simplefilter("always", self.expected)
+        if isinstance(self.expected, tuple):
```
Comment (DanielNoord) [inline Lib/unittest/case.py:277]: This should probably be done in a separate PR, but `issubclass` does not support a tuple. However, the type of the `category` parameter of `warnings.simplefilter` doesn't really indicate that it should support a tuple anyway.
I could also change it so that the `issubclass` checks for `tuples` before doing the check, but I'm not sure if that is actually correct.

Let me know!

**Review by MaxwellDupre — APPROVED** [review]
Ran 117 tests in 2.152s
OK
Looks ok to me.

**Comment by DanielNoord** [comment]
@berkerpeksag Gentle ping as this is basically your patch from https://bugs.python.org/issue16845 > https://bugs.python.org/file35773/issue16845.diff.

Sorry if such pings are not appreciated!

File: Lib/unittest/case.py (hunk 1 above)
Comment (serhiy-storchaka) [reply Lib/unittest/case.py:277]: I think that there may be a third-party code which uses `simplefilter()` with a tuple. Instead of changing unittest, it is better to make the argument check in `simplefilter()` more lenient.

It would be nice to make also `filterwarnings()` accepting a tuple.

**Comment by DanielNoord** [comment]
I'm closing this PR. On master `warnings.simplefilter("error", int)` works fine so I don't see a reason to add stricter validation here.

@serhiy-storchaka if you agree do you mind closing the associated issue?
